service/controller/v3/resource/instance/create.go

```diff
@@ -33,187 +49,205 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
-			if err != nil {
-				return microerror.Mask(err)
-			}
-			err = r.createDrainerConfig(ctx, customObject, masterInstanceToDrain, key.MasterInstanceName)
-			if err != nil {
-				return microerror.Mask(err)
+	if hasResourceStatus(customObject, Stage, InstancesUpgrading) {
+		versionValue := map[string]string{}
+		{
+			for _, node := range customObject.Status.Cluster.Nodes {
+				versionValue[node.Name] = node.Version
 			}
-			err = r.reimageInstance(ctx, customObject, masterInstanceToReimage, key.MasterVMSSName, key.MasterInstanceName)
-			if err != nil {
-				return microerror.Mask(err)
+		}
+
+		var nodeConfigs []corev1alpha1.DrainerConfig
+		{
+			n := v1.NamespaceAll
+			o := metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("%s=%s", key.ClusterIDLabel, key.ClusterID(customObject)),
 			}
-			err = r.deleteDrainerConfig(ctx, customObject, masterInstanceToReimage, key.MasterInstanceName, nodeConfigs)
+
+			list, err := r.g8sClient.CoreV1alpha1().DrainerConfigs(n).List(o)
 			if err != nil {
 				return microerror.Mask(err)
 			}
 
-			r.logger.LogCtx(ctx, "level", "debug", "message", "processed master VMSSs")
+			nodeConfigs = list.Items
 		}
-	}
 
-	// In case the master instance is being updated we want to prevent any
-	// other updates on the workers. This is because the update process
-	// involves the draining of the updated node and if the master is being
-	// updated at the same time the guest cluster's Kubernetes API is not
-	// available in order to drain nodes. As consequence we have to reset the
-	// worker instance selected to be reimaged in order to not update its
-	// version information. The next reconciliation loop will catch up here
-	// and instruct the worker instance to be reimaged again.
-	if masterInstanceToDrain == nil && masterInstanceToReimage == nil {
-		allWorkerInstances, err := r.allInstances(ctx, customObject, key.WorkerVMSSName)
-		if IsScaleSetNotFound(err) {
-			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find the scale set '%s'", key.WorkerVMSSName(customObject)))
-		} else if err != nil {
-			return microerror.Mask(err)
-		} else {
-			r.logger.LogCtx(ctx, "level", "debug", "message", "processing worker VMSSs")
-
-			workerInstanceToUpdate, workerInstanceToDrain, workerInstanceToReimage, err := r.nextInstance(ctx, customObject, allWorkerInstances, nodeConfigs, key.WorkerInstanceName, workerVersionsValue)
-			if err != nil {
-				return microerror.Mask(err)
-			}
-			err = r.updateInstance(ctx, customObject, workerInstanceToUpdate, key.WorkerVMSSName, key.WorkerInstanceName)
-			if err != nil {
-				return microerror.Mask(err)
-			}
-			err = r.createDrainerConfig(ctx, customObject, workerInstanceToDrain, key.WorkerInstanceName)
-			if err != nil {
-				return microerror.Mask(err)
-			}
-			err = r.reimageInstance(ctx, customObject, workerInstanceToReimage, key.WorkerVMSSName, key.WorkerInstanceName)
-			if err != nil {
+		var masterUpgraded bool
```

Comment:
The whole block about upgrading master is repeated for workers, it would be great to move this into a function to avoid repetition.

Reply:
You have a good point but I am not addressing it in this PR because it was like this already and the PR already derailed more than I anticipated. For sure something to boyscout down the road.